plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java

```diff
@@ -0,0 +1,454 @@
+
+    private final ThreadPool threadPool;
+    private final TransportSearchAction searchAction;
+    private final TransportSearchScrollAction scrollAction;
+    private final Client client;
+
+    @Inject
+    protected TransportDeleteByQueryAction(Settings settings, ThreadPool threadPool, Client client,
+                                           TransportSearchAction transportSearchAction,
+                                           TransportSearchScrollAction transportSearchScrollAction,
+                                           TransportService transportService, ActionFilters actionFilters) {
+        super(settings, DeleteByQueryAction.NAME, threadPool, transportService, actionFilters, DeleteByQueryRequest.class);
+        this.threadPool = threadPool;
+        this.searchAction = transportSearchAction;
+        this.scrollAction = transportSearchScrollAction;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+        new AsyncDeleteByQueryAction(request, listener).start();
+    }
+
+    class AsyncDeleteByQueryAction {
+
+        private final DeleteByQueryRequest request;
+        private final ActionListener<DeleteByQueryResponse> listener;
+
+        private final long startTime;
+        private final AtomicBoolean timedOut;
+        private final AtomicLong total;
+
+        private final AtomicInteger bulks;
+        private final Semaphore semaphore;
+
+        private volatile ShardOperationFailedException[] shardFailures;
+        private final CopyOnWriteArrayList<IndexDeleteByQueryResponse> bulkResults;
+
+        AsyncDeleteByQueryAction(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+            this.request = request;
+            this.listener = listener;
+            this.startTime = System.currentTimeMillis();
+            this.timedOut = new AtomicBoolean(false);
+            this.total = new AtomicLong(0L);
+            this.semaphore = new Semaphore(0);
+            this.bulks = new AtomicInteger(0);
+            this.shardFailures = ShardSearchFailure.EMPTY_ARRAY;
+            this.bulkResults = new CopyOnWriteArrayList<>();
+        }
+
+        public void start() {
+            executeScan();
+        }
+
+        void executeScan() {
+            try {
+                final SearchRequest scanRequest = new SearchRequest(request.indices()).types(request.types()).indicesOptions(request.indicesOptions());
+                scanRequest.searchType(SearchType.SCAN).scroll(request.scroll());
+                if (request.routing() != null) {
+                    scanRequest.routing(request.routing());
+                }
+
+                SearchSourceBuilder source = new SearchSourceBuilder().query(request.source()).fields("_routing", "_parent").fetchSource(false).version(true);
```

Comment:
Can we check if we need to get `_parent` as well? I can't recall if we use it as routing without storing it in the _routing field, I think we do always also put it in _routing, need to check